File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/mutation/SemanticMutationBuilder.java
package com.foundgine.core.semantic.mutation;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.query.SemanticFilterExpression;
import java.util.*;

/** Factory for constructing canonical semantic mutation operations. */
public final class SemanticMutationBuilder {
	private SemanticMutationBuilder() {
	}

	public static SemanticMutationOperation create(EntityId entity, List<SemanticMutationField> fields,
			List<FieldId> returnFields) {
		return operation(entity, SemanticMutationKind.CREATE, fields, null, List.of(), returnFields);
	}

	public static SemanticMutationOperation update(EntityId entity, List<SemanticMutationField> fields,
			SemanticFilterExpression filter, List<FieldId> returnFields) {
		return operation(entity, SemanticMutationKind.UPDATE, fields, filter, List.of(), returnFields);
	}

	public static SemanticMutationOperation delete(EntityId entity, SemanticFilterExpression filter,
			List<FieldId> returnFields) {
		return operation(entity, SemanticMutationKind.DELETE, List.of(), filter, List.of(), returnFields);
	}

	public static SemanticMutationOperation upsert(EntityId entity, List<SemanticMutationField> fields,
			List<FieldId> conflictFields, List<FieldId> returnFields) {
		return operation(entity, SemanticMutationKind.UPSERT, fields, null, conflictFields, returnFields);
	}

	private static SemanticMutationOperation operation(EntityId entity, SemanticMutationKind kind,
			List<SemanticMutationField> fields, SemanticFilterExpression filter, List<FieldId> conflicts,
			List<FieldId> returns) {
		var fs = fields == null ? List.<SemanticMutationField>of() : List.copyOf(fields);
		var effects = new ArrayList<SemanticMutationEffect>();
		var effectKind = switch (kind) {
		case CREATE -> SemanticMutationEffectKind.CREATE_ENTITY;
		case UPDATE -> SemanticMutationEffectKind.UPDATE_ENTITY;
		case DELETE -> SemanticMutationEffectKind.DELETE_ENTITY;
		case UPSERT -> SemanticMutationEffectKind.UPSERT_ENTITY;
		};
		effects.add(new SemanticMutationEffect(effectKind, entity));
		fs.forEach(
				f -> effects.add(new SemanticMutationEffect(SemanticMutationEffectKind.SET_FIELD, entity, f.field())));
		return new SemanticMutationOperation(entity, kind, fs, filter,
				conflicts == null ? List.of() : List.copyOf(conflicts),
				returns == null ? List.of() : List.copyOf(returns), effects, List.of());
	}
}

File: src/csharp/Foundgine.Core/Semantic/Mutation/SemanticMutationBuilder.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Mutation;

/// <summary>
/// Small factory for constructing canonical semantic mutation operations.
/// Validation of schema-specific legality remains a planner concern.
/// </summary>
public static class SemanticMutationBuilder
{
    public static SemanticMutationOperation Create(
        EntityId entity,
        IReadOnlyList<SemanticMutationField> fields,
        IReadOnlyList<FieldId>? returnFields = null) =>
        new(
            entity,
            SemanticMutationKind.Create,
            fields,
            null,
            Array.Empty<FieldId>(),
            returnFields ?? Array.Empty<FieldId>(),
            BuildEffects(entity, SemanticMutationKind.Create, fields),
            Array.Empty<SemanticMutationDependency>());

    public static SemanticMutationOperation Update(
        EntityId entity,
        IReadOnlyList<SemanticMutationField> fields,
        Foundgine.Core.Semantic.Query.SemanticFilterExpression? filter = null,
        IReadOnlyList<FieldId>? returnFields = null) =>
        new(
            entity,
            SemanticMutationKind.Update,
            fields,
            filter,
            Array.Empty<FieldId>(),
            returnFields ?? Array.Empty<FieldId>(),
            BuildEffects(entity, SemanticMutationKind.Update, fields),
            Array.Empty<SemanticMutationDependency>());

    public static SemanticMutationOperation Delete(
        EntityId entity,
        Foundgine.Core.Semantic.Query.SemanticFilterExpression filter,
        IReadOnlyList<FieldId>? returnFields = null) =>
        new(
            entity,
            SemanticMutationKind.Delete,
            Array.Empty<SemanticMutationField>(),
            filter,
            Array.Empty<FieldId>(),
            returnFields ?? Array.Empty<FieldId>(),
            BuildEffects(entity, SemanticMutationKind.Delete, Array.Empty<SemanticMutationField>()),
            Array.Empty<SemanticMutationDependency>());

    public static SemanticMutationOperation Upsert(
        EntityId entity,
        IReadOnlyList<SemanticMutationField> fields,
        IReadOnlyList<FieldId> conflictFields,
        IReadOnlyList<FieldId>? returnFields = null) =>
        new(
            entity,
            SemanticMutationKind.Upsert,
            fields,
            null,
            conflictFields,
            returnFields ?? Array.Empty<FieldId>(),
            BuildEffects(entity, SemanticMutationKind.Upsert, fields),
            Array.Empty<SemanticMutationDependency>());

    private static IReadOnlyList<SemanticMutationEffect> BuildEffects(
        EntityId entity,
        SemanticMutationKind kind,
        IReadOnlyList<SemanticMutationField> fields)
    {
        var effects = new List<SemanticMutationEffect>
        {
            new(
                kind switch
                {
                    SemanticMutationKind.Create => SemanticMutationEffectKind.CreateEntity,
                    SemanticMutationKind.Update => SemanticMutationEffectKind.UpdateEntity,
                    SemanticMutationKind.Delete => SemanticMutationEffectKind.DeleteEntity,
                    SemanticMutationKind.Upsert => SemanticMutationEffectKind.UpsertEntity,
                    _ => throw new ArgumentOutOfRangeException(nameof(kind))
                },
                entity)
        };

        effects.AddRange(fields.Select(x =>
            new SemanticMutationEffect(
                SemanticMutationEffectKind.SetField,
                entity,
                x.Field)));

        return effects;
    }
}
